modulemd_tools/modulemd_tools/yaml.py

```diff
@@ -0,0 +1,183 @@
+    # dependencies one be one and dropping all of them at once.
+    # We need to help ourselves a little and drop all runtime dependencies and
+    # re-populate them with the old ones if a new ones weren't set. Similarly
+    # for buildrequires.
+    old_deps = mod_stream.get_dependencies()[0]
+    new_deps = Modulemd.Dependencies()
+    if requires:
+        for depname, depstreams in requires.items():
+            for depstream in depstreams:
+                new_deps.add_runtime_stream(depname, depstream)
+    else:
+        for depname in old_deps.get_runtime_modules():
+            for depstream in old_deps.get_runtime_streams(depname):
+                new_deps.add_runtime_stream(depname, depstream)
+
+    if buildrequires:
+        for depname, depstreams in buildrequires.items():
+            for depstream in depstreams:
+                new_deps.add_buildtime_stream(depname, depstream)
+    else:
+        for depname in old_deps.get_buildtime_modules():
+            for depstream in old_deps.get_buildtime_streams(depname):
+                new_deps.add_buildtime_stream(depname, depstream)
+
+    if requires or buildrequires:
+        for dep in mod_stream.get_dependencies():
+            mod_stream.remove_dependencies(dep)
+        mod_stream.add_dependencies(new_deps)
+
+    idx2 = Modulemd.ModuleIndex.new()
+    idx2.add_module_stream(mod_stream)
+    return idx2.dump_to_string()
+
+
+def load(path):
+    """
+    Load modulemd YAML definition from a file
+    """
+    pass
+
+
+def dump(yaml, dest=None):
+    """
+    Dump modulemd YAML definition into a file
+
+    If `dest` is not specified, the file will be created in the current working
+    directory and it's name is going to be generated from the module attributes
+    in the `N:S:V:C:A.modulemd.yaml` format.
+
+    If `dest` is a directory, then the file is going to be generated in that
+    directory.
+
+    If `dest` is pointing to a (nonexisting) file, it is going to be dumped in
+    its place.
+    """
+    filename = _generate_filename(yaml)
+
+    path = dest
+    if not path:
+        path = os.path.join(os.getcwd(), filename)
+    elif os.path.isdir(path):
+        path = os.path.join(dest, filename)
+
+    with open(path, "w") as moduleyaml:
+        moduleyaml.write(yaml)
+
+
+def _generate_filename(yaml):
+    """
+    Generate filename for a module yaml
+    """
+    mod_stream = _yaml2stream(yaml)
+
+    return "{N}:{S}:{V}:{C}:{A}.modulemd.yaml".format(
+        N=mod_stream.get_module_name(), S=mod_stream.get_stream_name(),
+        V=mod_stream.get_version(), C=mod_stream.get_context(),
+        A=mod_stream.get_arch() or "noarch")
+
+
+def _yaml2stream(yaml):
+    idx = Modulemd.ModuleIndex.new()
+    idx.update_from_string(yaml, strict=True)
+
+    mod_names = idx.get_module_names()
+    mod = idx.get_module(mod_names[0])
+    mod_stream = mod.get_all_streams()[0]
+    return mod_stream
```

Comment:
This function should throw an exception if the YAML it was passed results in an index containing more than one module or stream.